File: backend/src/services/log.rs
// moved category, entry, and tag to their own modules
// exporting from log for backwards compatibility
// but usage should migrate to the new modules
// #TODO: move usage, see above :)
pub use super::category::*;
pub use super::entry::*;
pub use super::tag::*;
use crate::{
  establish_connection,
  schema::{categories, entries, tags},
  services::get_user,
  util::EphemerideError,
};
use diesel::{ExpressionMethods, QueryDsl, RunQueryDsl};

pub fn create_default_data(user_id: String) -> Result<bool, EphemerideError> {
  let default_categories = vec!["Activities", "Tags"];

  let default_tags = vec![
    ("Activities", "Work", "base"),
    ("Activities", "Movie", "base"),
    ("Activities", "Exercise", "base"),
    ("Activities", "Read", "base"),
    ("Activities", "Shopping", "base"),
    ("Activities", "Gaming", "base"),
    ("Tags", "Travel", "base"),
    ("Tags", "Important", "blue"),
    ("Tags", "Sick", "red"),
  ];

  for category_name in default_categories {
    let category_result = create_category(CreateCategory {
      name: category_name.to_string(),
      user_id: user_id.clone(),
    });

    let category = match category_result {
      Ok(category) => category,
      Err(_) => return Err(EphemerideError::DatabaseError),
    };

    for (cat_name, tag_name, color) in &default_tags {
      if *cat_name == category_name {
        let tag_result = create_tag(CreateTag {
          name: tag_name.to_string(),
          color: color.to_string(),
          category_id: category.id.clone(),
          user_id: user_id.clone(),
        });

        if tag_result.is_err() {
          return Err(EphemerideError::DatabaseError);
        }
      }
    }
  }

  Ok(true)
}

pub fn delete_all_user_data(user_id: &str) -> Result<bool, EphemerideError> {
  let user = get_user(user_id);

  if user.is_err() {
    return Err(EphemerideError::UserNotFound);
  }

  let mut conn = establish_connection();

  let delete_entries =
    diesel::delete(entries::table.filter(entries::user_id.eq(user_id))).execute(&mut conn);

  if delete_entries.is_err() {
    return Err(EphemerideError::DatabaseError);
  }

  let delete_tags =
    diesel::delete(tags::table.filter(tags::user_id.eq(user_id))).execute(&mut conn);

  if delete_tags.is_err() {
    return Err(EphemerideError::DatabaseError);
  }

  let delete_categories =
    diesel::delete(categories::table.filter(categories::user_id.eq(user_id))).execute(&mut conn);

  if delete_categories.is_err() {
    return Err(EphemerideError::DatabaseError);
  }

  Ok(true)
}

File: frontend/src/lib/types/log.ts
import type { Color } from './color'
import type { MoodValue } from './components/moodinput'

export type Category = {
  id: string
  name: string
  user_id: string
  created_at: number
}

export type NewCategory = {
  name: string
}

export type EditCategory = {
  id: string
  name: string
}

export type CategoryWithTags = Category & {
  tags: Tag[]
  selected_tags?: string[]
}

export type Tag = {
  id: string
  user_id: string
  created_at: number
  name: string
  color: Color
  category_id: string
}

export type NewTag = {
  name: string
  color: Color
  category_id: string
}

export type EditTag = {
  id: string
  name: string
  color: Color
}

export type Entry = {
  id: string
  user_id: string
  date: string
  created_at: number
  mood: MoodValue
  entry?: string
  selected_tags: string[]
}

export type NewEntry = {
  date: string
  mood: MoodValue
  entry?: string
  selected_tags: string[]
}

export type EditEntry = {
  id: string
  date: string
  mood: MoodValue
  entry?: string
  selected_tags: string[]
}

export const entryMaxLength = 1000

/**
 * see /backend/src/services/log.rs
 */
export const defaultCategories: CategoryWithTags[] = [
  {
    id: crypto.randomUUID(),
    name: 'Activities',
    user_id: '',
    created_at: 0,
    tags: [
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Work',
        color: 'base',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Movie',
        color: 'base',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Exercise',
        color: 'base',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Read',
        color: 'base',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Shopping',
        color: 'base',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Gaming',
        color: 'base',
        category_id: '',
      },
    ],
  },
  {
    id: crypto.randomUUID(),
    name: 'Tags',
    user_id: '',
    created_at: 0,
    tags: [
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Travel',
        color: 'base',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Important',
        color: 'blue',
        category_id: '',
      },
      {
        id: crypto.randomUUID(),
        user_id: '',
        created_at: 0,
        name: 'Sick',
        color: 'red',
        category_id: '',
      },
    ],
  },
]
